### ISSUE TITLE
mempool: CListMempool.CheckTx misuses read-lock to protect write operations; possible data race

### ISSUE THREAD

**Issue by odeke-em** [cometbft/cometbft#2444]
## Bug Report
Noticed in an audit that there is a misuse of read-locks around mempool/CListMempool for example in #2443 and in particular for this bug report let's look at this code 
https://github.com/cometbft/cometbft/blob/4ce0277b35f31985bbf2c25d3806a184a4510010/mempool/clist_mempool.go#L251C12-L251-L254

but meanwhile in that code protected, we can see this invocation https://github.com/cometbft/cometbft/blob/4ce0277b35f31985bbf2c25d3806a184a4510010/mempool/clist_mempool.go#L280 and even later on
https://github.com/cometbft/cometbft/blob/4ce0277b35f31985bbf2c25d3806a184a4510010/mempool/clist_mempool.go#L290

because we aren't sure how guarantees that .CheckTxAsync won't be mutating.

**Comment by hvanz** [comment]
That's correct. If we are using the local ABCI client, `CheckTxAsync` will execute the callback function right after receiving the response from the app, possibly adding the transaction to the mempool while it still holds the lock. So `RLock` should be changed to `Lock`. On the other side, the cache has it's own mutex, and it's not supposed to be guarded by `updateMtx`.

**Comment by hoanguyenkh** [comment]
thank you, i have a PR here -> https://github.com/cometbft/cometbft/pull/2497

**Comment by melekes** [comment]
> So RLock should be changed to Lock

But the `updateMtx` is `Exclusive mutex for Update method to prevent concurrent execution of CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.`?

**Comment by melekes** [comment]
@hvanz, is this still actual, given the recent developments?
